In [0]:
from pyspark.sql.functions import (
    col, count, when, to_date, try_to_date, year, explode, split, regexp_replace, regexp_extract, sum, when
)


In [0]:
data = spark.table("workspace.default.steam_game_output")
data = data.toDF(*[c.replace(".", "_") for c in data.columns])
display(data.limit(10))

appid,categories,ccu,developer,discount,genre,header_image,id,initialprice,languages,name,negative,owners,platforms_linux,platforms_mac,platforms_windows,positive,price,publisher,release_date,required_age,short_description,tags_1980s,tags_1990's,tags_2_5D,tags_2D,tags_2D Fighter,tags_2D Platformer,tags_360 Video,tags_3D,tags_3D Fighter,tags_3D Platformer,tags_3D Vision,tags_4 Player Local,tags_4X,tags_6DOF,tags_8-bit Music,tags_ATV,tags_Abstract,tags_Action,tags_Action RPG,tags_Action RTS,tags_Action Roguelike,tags_Action-Adventure,tags_Addictive,tags_Adventure,tags_Agriculture,tags_Aliens,tags_Alternate History,tags_Ambient,tags_America,tags_Animation & Modeling,tags_Anime,tags_Arcade,tags_Archery,tags_Arena Shooter,tags_Artificial Intelligence,tags_Assassin,tags_Asymmetric VR,tags_Asynchronous Multiplayer,tags_Atmospheric,tags_Audio Production,tags_Auto Battler,tags_Automation,tags_Automobile Sim,tags_BMX,tags_Base-Building,tags_Baseball,tags_Based On A Novel,tags_Basketball,tags_Battle Royale,tags_Beat 'em up,tags_Beautiful,tags_Benchmark,tags_Bikes,tags_Blood,tags_Board Game,tags_Boss Rush,tags_Bowling,tags_Boxing,tags_Building,tags_Bullet Hell,tags_Bullet Time,tags_CRPG,tags_Capitalism,tags_Card Battler,tags_Card Game,tags_Cartoon,tags_Cartoony,tags_Casual,tags_Cats,tags_Character Action Game,tags_Character Customization,tags_Chess,tags_Choices Matter,tags_Choose Your Own Adventure,tags_Cinematic,tags_City Builder,tags_Class-Based,tags_Classic,tags_Clicker,tags_Co-op,tags_Co-op Campaign,tags_Coding,tags_Cold War,tags_Collectathon,tags_Colony Sim,tags_Colorful,tags_Combat,tags_Combat Racing,tags_Comedy,tags_Comic Book,tags_Competitive,tags_Conspiracy,tags_Controller,tags_Conversation,tags_Cooking,tags_Cozy,tags_Crafting,tags_Creature Collector,tags_Cricket,tags_Crime,tags_Crowdfunded,tags_Cult Classic,tags_Cute,tags_Cyberpunk,tags_Cycling,tags_Dark,tags_Dark Comedy,tags_Dark Fantasy,tags_Dark Humor,tags_Dating Sim,tags_Deckbuilding,tags_Demons,tags_Design & Illustration,tags_Destruction,tags_Detective,tags_Difficult,tags_Dinosaurs,tags_Diplomacy,tags_Documentary,tags_Dog,tags_Dragons,tags_Drama,tags_Driving,tags_Dungeon Crawler,tags_Dungeons & Dragons,tags_Dynamic Narration,tags_Dystopian,tags_Early Access,tags_Economy,tags_Education,tags_Electronic,tags_Electronic Music,tags_Emotional,tags_Epic,tags_Episodic,tags_Escape Room,tags_Experience,tags_Experimental,tags_Exploration,tags_FMV,tags_FPS,tags_Faith,tags_Family Friendly,tags_Fantasy,tags_Farming,tags_Farming Sim,tags_Fast-Paced,tags_Feature Film,tags_Female Protagonist,tags_Fighting,tags_First-Person,tags_Fishing,tags_Flight,tags_Football,tags_Foreign,tags_Free to Play,tags_Funny,tags_Futuristic,tags_Gambling,tags_Game Development,tags_GameMaker,tags_Games Workshop,tags_Gaming,tags_God Game,tags_Golf,tags_Gore,tags_Gothic,tags_Grand Strategy,tags_Great Soundtrack,tags_Grid-Based Movement,tags_Gun Customization,tags_Hack and Slash,tags_Hacking,tags_Hand-drawn,tags_Hardware,tags_Heist,tags_Hentai,tags_Hero Shooter,tags_Hex Grid,tags_Hidden Object,tags_Historical,tags_Hockey,tags_Horror,tags_Horses,tags_Hunting,tags_Idler,tags_Illuminati,tags_Immersive,tags_Immersive Sim,tags_Indie,tags_Instrumental Music,tags_Intentionally Awkward Controls,tags_Interactive Fiction,tags_Inventory Management,tags_Investigation,tags_Isometric,tags_JRPG,tags_Jet,tags_Job Simulator,tags_Jump Scare,tags_Kickstarter,tags_LEGO,tags_LGBTQ+,tags_Lemmings,tags_Level Editor,tags_Life Sim,tags_Linear,tags_Local Co-Op,tags_Local Multiplayer,tags_Logic,tags_Loot,tags_Looter Shooter,tags_Lore-Rich,tags_Lovecraftian,tags_MMORPG,tags_MOBA,tags_Magic,tags_Mahjong,tags_Management,tags_Mars,tags_Martial Arts,tags_Massively Multiplayer,tags_Masterpiece,tags_Match 3,tags_Mature,tags_Mechs,tags_Medical Sim,tags_Medieval,tags_Memes,tags_Metroidvania,tags_Military,tags_Mini Golf,tags_Minigames,tags_Minimalist,tags_Mining,tags_Mod,tags_Moddable,tags_Modern,tags_Motocross,tags_Motorbike,tags_Mouse only,tags_Movie,

In [0]:
data.printSchema()

root
 |-- appid: long (nullable = true)
 |-- categories: array (nullable = true)
 |    |-- element: string (containsNull = true)
 |-- ccu: long (nullable = true)
 |-- developer: string (nullable = true)
 |-- discount: string (nullable = true)
 |-- genre: string (nullable = true)
 |-- header_image: string (nullable = true)
 |-- id: long (nullable = true)
 |-- initialprice: string (nullable = true)
 |-- languages: string (nullable = true)
 |-- name: string (nullable = true)
 |-- negative: long (nullable = true)
 |-- owners: string (nullable = true)
 |-- platforms_linux: boolean (nullable = true)
 |-- platforms_mac: boolean (nullable = true)
 |-- platforms_windows: boolean (nullable = true)
 |-- positive: long (nullable = true)
 |-- price: string (nullable = true)
 |-- publisher: string (nullable = true)
 |-- release_date: string (nullable = true)
 |-- required_age: string (nullable = true)
 |-- short_description: string (nullable = true)
 |-- tags_1980s: double (nullable = true)
 |-- tag

In [0]:
(data.count(), len(data.columns))

(55691, 465)

In [0]:
missing_counts = data.select([
    sum(when(col(f"`{c}`").isNull(), 1).otherwise(0)).alias(c)
    for c in data.columns
])

display(missing_counts)

appid,categories,ccu,developer,discount,genre,header_image,id,initialprice,languages,name,negative,owners,platforms_linux,platforms_mac,platforms_windows,positive,price,publisher,release_date,required_age,short_description,tags_1980s,tags_1990's,tags_2_5D,tags_2D,tags_2D Fighter,tags_2D Platformer,tags_360 Video,tags_3D,tags_3D Fighter,tags_3D Platformer,tags_3D Vision,tags_4 Player Local,tags_4X,tags_6DOF,tags_8-bit Music,tags_ATV,tags_Abstract,tags_Action,tags_Action RPG,tags_Action RTS,tags_Action Roguelike,tags_Action-Adventure,tags_Addictive,tags_Adventure,tags_Agriculture,tags_Aliens,tags_Alternate History,tags_Ambient,tags_America,tags_Animation & Modeling,tags_Anime,tags_Arcade,tags_Archery,tags_Arena Shooter,tags_Artificial Intelligence,tags_Assassin,tags_Asymmetric VR,tags_Asynchronous Multiplayer,tags_Atmospheric,tags_Audio Production,tags_Auto Battler,tags_Automation,tags_Automobile Sim,tags_BMX,tags_Base-Building,tags_Baseball,tags_Based On A Novel,tags_Basketball,tags_Battle Royale,tags_Beat 'em up,tags_Beautiful,tags_Benchmark,tags_Bikes,tags_Blood,tags_Board Game,tags_Boss Rush,tags_Bowling,tags_Boxing,tags_Building,tags_Bullet Hell,tags_Bullet Time,tags_CRPG,tags_Capitalism,tags_Card Battler,tags_Card Game,tags_Cartoon,tags_Cartoony,tags_Casual,tags_Cats,tags_Character Action Game,tags_Character Customization,tags_Chess,tags_Choices Matter,tags_Choose Your Own Adventure,tags_Cinematic,tags_City Builder,tags_Class-Based,tags_Classic,tags_Clicker,tags_Co-op,tags_Co-op Campaign,tags_Coding,tags_Cold War,tags_Collectathon,tags_Colony Sim,tags_Colorful,tags_Combat,tags_Combat Racing,tags_Comedy,tags_Comic Book,tags_Competitive,tags_Conspiracy,tags_Controller,tags_Conversation,tags_Cooking,tags_Cozy,tags_Crafting,tags_Creature Collector,tags_Cricket,tags_Crime,tags_Crowdfunded,tags_Cult Classic,tags_Cute,tags_Cyberpunk,tags_Cycling,tags_Dark,tags_Dark Comedy,tags_Dark Fantasy,tags_Dark Humor,tags_Dating Sim,tags_Deckbuilding,tags_Demons,tags_Design & Illustration,tags_Destruction,tags_Detective,tags_Difficult,tags_Dinosaurs,tags_Diplomacy,tags_Documentary,tags_Dog,tags_Dragons,tags_Drama,tags_Driving,tags_Dungeon Crawler,tags_Dungeons & Dragons,tags_Dynamic Narration,tags_Dystopian,tags_Early Access,tags_Economy,tags_Education,tags_Electronic,tags_Electronic Music,tags_Emotional,tags_Epic,tags_Episodic,tags_Escape Room,tags_Experience,tags_Experimental,tags_Exploration,tags_FMV,tags_FPS,tags_Faith,tags_Family Friendly,tags_Fantasy,tags_Farming,tags_Farming Sim,tags_Fast-Paced,tags_Feature Film,tags_Female Protagonist,tags_Fighting,tags_First-Person,tags_Fishing,tags_Flight,tags_Football,tags_Foreign,tags_Free to Play,tags_Funny,tags_Futuristic,tags_Gambling,tags_Game Development,tags_GameMaker,tags_Games Workshop,tags_Gaming,tags_God Game,tags_Golf,tags_Gore,tags_Gothic,tags_Grand Strategy,tags_Great Soundtrack,tags_Grid-Based Movement,tags_Gun Customization,tags_Hack and Slash,tags_Hacking,tags_Hand-drawn,tags_Hardware,tags_Heist,tags_Hentai,tags_Hero Shooter,tags_Hex Grid,tags_Hidden Object,tags_Historical,tags_Hockey,tags_Horror,tags_Horses,tags_Hunting,tags_Idler,tags_Illuminati,tags_Immersive,tags_Immersive Sim,tags_Indie,tags_Instrumental Music,tags_Intentionally Awkward Controls,tags_Interactive Fiction,tags_Inventory Management,tags_Investigation,tags_Isometric,tags_JRPG,tags_Jet,tags_Job Simulator,tags_Jump Scare,tags_Kickstarter,tags_LEGO,tags_LGBTQ+,tags_Lemmings,tags_Level Editor,tags_Life Sim,tags_Linear,tags_Local Co-Op,tags_Local Multiplayer,tags_Logic,tags_Loot,tags_Looter Shooter,tags_Lore-Rich,tags_Lovecraftian,tags_MMORPG,tags_MOBA,tags_Magic,tags_Mahjong,tags_Management,tags_Mars,tags_Martial Arts,tags_Massively Multiplayer,tags_Masterpiece,tags_Match 3,tags_Mature,tags_Mechs,tags_Medical Sim,tags_Medieval,tags_Memes,tags_Metroidvania,tags_Military,tags_Mini Golf,tags_Minigames,tags_Minimalist,tags_Mining,tags_Mod,tags_Moddable,tags_Modern,tags_Motocross,tags_Motorbike,tags_Mouse only,tags_Movie,

# General Analysis of data distribution

## Which publisher has released the most games on Steam?

In [0]:
top_publishers = (
    data
    .groupBy("publisher")
    .count()
    .orderBy("count", ascending=False)
    .limit(10)
)

display(top_publishers)

publisher,count
Big Fish Games,422
8floor,202
SEGA,165
Strategy First,151
Square Enix,141
Choice of Games,140
HH-Games,132
Sekai Project,132
,132
Ubisoft,127


Databricks visualization. Run in Databricks to view.

#### Big Fish Games had published more games than any other name of the industry listed in this dataset.

In [0]:
top_publishers.write.mode("overwrite").saveAsTable("top_publisher")

## What are the best rated games?


In [0]:
top_games = (
    data
    .select("name", "positive")
    .orderBy(col("positive").desc())
    .limit(10)
)

display(top_games)

name,positive
Counter-Strike: Global Offensive,5943345
Dota 2,1534895
Grand Theft Auto V,1229265
PUBG: BATTLEGROUNDS,1185361
Terraria,1014711
Tom Clancy's Rainbow Six Siege,942910
Garry's Mod,861240
Team Fortress 2,846407
Rust,732513
Left 4 Dead 2,643836


Databricks visualization. Run in Databricks to view.

#### The best rated games were Counter Strike: Global Offensive; Dota 2, and Grand Theft Auto 5

In [0]:
top_games.write.mode("overwrite").saveAsTable("top_games")

## Are there years with more releases? Were there more or fewer game releases during the Covid, for example?

In [0]:
df_years = (
    data
    .withColumn("release_date", try_to_date(col("release_date"), "yyyy/M/d"))
    .withColumn("release_year", year(col("release_date")))
    .groupBy("release_year")
    .count()
    .orderBy("release_year")
)

display(df_years)

release_year,count
null,222
1997,2
1998,1
1999,3
2000,2
2001,4
2002,1
2003,3
2004,6
2005,6


In [0]:
df_years_clean = (
    data
    .withColumn("release_date_clean", try_to_date(col("release_date"), "yyyy/M/d"))
    .filter(col("release_date_clean").isNotNull())
    .withColumn("release_year", year(col("release_date_clean")))
    .groupBy("release_year")
    .count()
    .orderBy("release_year")
)

display(df_years_clean)

release_year,count
1997,2
1998,1
1999,3
2000,2
2001,4
2002,1
2003,3
2004,6
2005,6
2006,61


Databricks visualization. Run in Databricks to view.

#### most releases happened after 2015 and peaked during the covid in 2020-2021

In [0]:
df_years_clean.write.mode("overwrite").saveAsTable("df_years")

## How are the prizes distributed? Are there many games with a discount?


In [0]:
from pyspark.sql.functions import round
df_price = (
    data
    .withColumn("price", col("price").cast("double"))
    .filter(col("price") > 1)
)
df_price = df_price.withColumn("price", (col("price") /100 ))
df_price = df_price.select("price")
display(df_price)

price
9.99
9.99
5.99
19.99
1.99
7.99
12.99
2.99
13.99
0.99


Databricks visualization. Run in Databricks to view.

In [0]:
df_price.write.mode("overwrite").saveAsTable("df_price")

In [0]:
df_discount = (
    data
    .withColumn("has_discount", when(col("discount") > 0, 'yes').otherwise('no'))
)
df_discount = df_discount.select("appid", "has_discount")
display(
    df_discount.groupBy("has_discount").count()
)

has_discount,count
no,53173
yes,2518


Databricks visualization. Run in Databricks to view.

#### After visualization it was found that very few discounts were in this dataset.

In [0]:
df_discount.write.mode("overwrite").saveAsTable("df_discount")

## What are the most represented languages?


In [0]:
df_lang = (
    data
    .select(explode(split(col("languages"), ";")).alias("language"))
    .groupBy("language")
    .count()
    .orderBy("count", ascending=False)
    .limit(10)
)

display(df_lang)

language,count
English,29163
"English, Russian",1881
"English, Simplified Chinese",1200
"English, Japanese",1015
"English, Not supported, Simplified Chinese",683
"English, German",605
"English, French",510
"English, French, Italian, German, Spanish - Spain",493
"English, Portuguese - Brazil",463
"English, Simplified Chinese, Traditional Chinese",326


Databricks visualization. Run in Databricks to view.

#### English is used widely and internationally, it is of not surprise that we find it overwhelmingly available in games.

In [0]:
df_lang.write.mode("overwrite").saveAsTable("df_lang")

## Are there many games prohibited for children under 16/18?

In [0]:
df_age = (
    data
    .groupBy("required_age")
    .count()
    .orderBy("count", ascending=False)
    .limit(10)
)

display(df_age)

required_age,count
0,55030
15,264
18,223
17,38
16,38
12,32
13,26
14,10
10,7
180,4


Databricks visualization. Run in Databricks to view.

In [0]:

df_clean = data.withColumn(
    "required_age_clean",
    regexp_extract(col("required_age"), r"(\d+)", 1).cast("int")
)

In [0]:
df_age2 = df_clean.withColumn(
    "age_group",
    when(col("required_age_clean") == 0, "No restriction (0)")
    .when((col("required_age_clean") >= 16) & (col("required_age_clean") < 18), "16-17")
    .when(col("required_age_clean") >= 18, "18+")
    .otherwise("Unknown")
)

display(df_age2.groupBy("age_group").count())

age_group,count
No restriction (0),55030
Unknown,355
18+,230
16-17,76


In [0]:
age_dist = df_age2.groupBy("age_group").count().orderBy("count", ascending=False)
total = df_age2.count()

age_prop = (
    age_dist
    .withColumn("percentage", col("count") / total * 100)
)

display(age_prop)

age_group,count,percentage
No restriction (0),55030,98.81309367761398
Unknown,355,0.6374459068790289
18+,230,0.41299312276669475
16-17,76,0.13646729274029915


Databricks visualization. Run in Databricks to view.

#### Only ~.54 % of games were made for youth between 16 and 18 years old while ~98.81 % had no age restriction.
#### It sounds economically benefitial for a company to release their  games for any age as it widens the audience

In [0]:
age_prop.write.mode("overwrite").saveAsTable("age_prop")